File: C++/task5/part 2/linked_list.cpp
#include <iostream>
#include <string>
#include "linked_list.hpp"

void linked_list::push(node* newNode) {
    if (head == nullptr && tail == nullptr) {
        head = newNode;
        tail = newNode;
    } else {
        newNode -> next = head;
        newNode -> previous = nullptr;
        head -> previous = newNode;
        head = newNode;
    }
}

void linked_list::append(node* newNode) {
    if (head == nullptr && tail == nullptr) {
        head = newNode;
        tail = newNode;
    } else {
        newNode -> next = nullptr;
        newNode -> previous = tail;
        tail -> next = newNode;
        tail = newNode;
    }
}

bool linked_list::isIn(std::string& word) {
    node* currentNode = head;
    while (currentNode != nullptr) {
        if (currentNode -> data == word) {
            return true;
        }
        currentNode = currentNode -> next;
    }
    return false;
}

int linked_list::size() {
    int count = 0;
    node* currentNode = head;
    while (currentNode != nullptr) {
        count++;
        currentNode = currentNode -> next;
    }
    return count;
}

void linked_list::swap(node& a, node& b) {
    auto tmpData = b.data;
    b.data = a.data;
    a.data = tmpData;
}

void linked_list::sort() {
    bool sorted = false;
    node* currentNode = head;

    while (!sorted) {
        sorted = true;
        while (currentNode != nullptr) {
            if (currentNode -> next != nullptr && currentNode -> data > currentNode -> next -> data) {
                sorted = false;
                swap(*currentNode, *currentNode -> next);
            }
            currentNode = currentNode -> next;
        }
        currentNode = head;
    }
}

std::ostream& operator<<(std::ostream& os, linked_list& list) {
    node* currentNode = list.head;
    while (currentNode != nullptr) {
        if (currentNode == list.head) {
            os << "[" << currentNode -> data << ", ";
        } else if (currentNode -> next == nullptr) {
            os << currentNode -> data << "]";
        } else {
            os << currentNode -> data << ", ";
        }
        currentNode = currentNode -> next;
    }
    return os;
}

File: Python/task5/part 2/linked_list.py
#!/usr/bin/python3

from node import node

class linked_list:
    def __init__(self):
        self.size = 0
        self.first_node = None
        self.last_node = None

    def push(self, new_node):
        if self.first_node == None and self.last_node == None:
            self.first_node = new_node
            self.last_node = new_node
        else:
            new_node.next_node = self.first_node
            new_node.previous_node = None
            self.first_node.previous_node = new_node
            self.first_node = new_node
        self.size += 1

    def pop(self):
        node = self.last_node
        self.remove(node)
        return node

    def isIn(self, word):
        currentNode = self.first_node
        while currentNode:
            if currentNode.data == word:
                return True
            currentNode = currentNode.next_node
        return False

    def append(self, new_node):
        if self.first_node == None and self.last_node == None:
            self.first_node = new_node
            self.last_node = new_node
        else:
            new_node.next_node = None
            new_node.previous_node = self.last_node
            self.last_node.next_node = new_node
            self.last_node = new_node
        self.size += 1

    def remove(self, node):
        if node == self.first_node:
            self.first_node = node.next_node
        elif node == self.last_node:
            self.last_node = node.previous_node
            self.last_node.next_node = None
        else:
            next_node = node.next_node
            previous_node = node.previous_node
            next_node.previous_node = previous_node
            previous_node.next_node = next_node
        self.size -= 1

    def length(self):
        return self.size

    def swap(self, a, b):
        tmpData = b.data
        b.data = a.data
        a.data = tmpData

    def sort(self):
        currentNode = self.first_node
        Sorted = False

        while not Sorted:
            Sorted = True
            while currentNode:
                if currentNode.next_node != None and currentNode.data > currentNode.next_node.data:
                    self.swap(currentNode, currentNode.next_node)
                    Sorted = False
                currentNode = currentNode.next_node
            currentNode = self.first_node

    def reverse_display(self):
        contents = '['
        next_node = self.last_node
        while next_node:
            if next_node.previous_node == None:
                contents += '{0}]'.format(str(next_node.data))
            else:
                contents += '{0}, '.format(str(next_node.data))
            next_node = next_node.previous_node
        return contents

    def __str__(self):
        contents = '['
        next_node = self.first_node
        while next_node:
            if next_node.next_node == None:
                contents += '{0}]'.format(str(next_node.data))
            else:
                contents += '{0}, '.format(str(next_node.data))
            next_node = next_node.next_node
        return contents